background.js

```diff
@@ -913,14 +913,12 @@ const SendLater = {
 
           // Re-save message (drops x-send-later headers by default
           // because they are not loaded when editing as draft).
-          messenger.compose.saveMessage(tab.id, {mode: "draft"}).then(
-            () => {
-              SLTools.scheduledMsgCache.delete(originalMsg.id);
-              SLTools.unscheduledMsgCache.add(originalMsg.id);
-              SendLater.updateStatusIndicator();
-            }
-          );
-
+          // TODO: Not using the return value of saveMessage() ???
```

Comment:
The returned saved message is not used. It looked strange to me, any insights?

Reply:
I don't follow your question entirely.
The assumption of the code here is that the message ID of the newly saved version of the message is identical to the message ID of the previous incarnation of the message. Does that assumption hold true?
If so, then I don't see how it matters that the saved draft is not "used." The point here is to save the message to Drafts for later delivering at the scheduled time. Once the save happens this part of the code no longer has anything it needs to do with the message. Note that the code that does the background send scans all drafts to look for messages to send, so the hand-off occurs by virtue of putting the message into the Drafts folder.
Am I misunderstanding your question?

Reply:
Ok, if the id of the original draft is not used at all after this call, we are good. But the next two lines re-schedule with the original draft id, not with the saved new draft id, which *could* be different. A check on the ID will tell you that.

I am just trying to point out edge cases, which may cause some of the observed behaviours.

Reply:
That's the question I'm asking above. I think @jph-sendlater believed when he wrote this code that the `id` field he's using here is the message ID field, and that it will be the same in the saved draft as it was in the previous one. Is that wrong, then yeah, I think this is a bug that needs to be fixed.

Reply:
I will try to address this.

Reply:
I need help, I think this piece of code has multiple issues.

```
      let originalMsg = await messenger.SL3U.findAssociatedDraft(window.id).then(
        m => m ? messenger.messages.getFull(m.id) : null
      );
      if (originalMsg) {
        SLTools.scheduledMsgCache.delete(originalMsg.id);
        SLTools.unscheduledMsgCache.add(originalMsg.id);

        // Check if original message has x-send-later headers
        if (originalMsg.headers.hasOwnProperty("x-send-later-at")) {
          let { preferences, scheduleCache } = await messenger.storage.local.get(
            { preferences: {}, scheduleCache: {} }
          );

          // Re-save message (drops x-send-later headers by default
          // because they are not loaded when editing as draft).
          messenger.compose.saveMessage(tab.id, {mode: "draft"}).then(
            () => {
              SLTools.scheduledMsgCache.delete(originalMsg.id);
              SLTools.unscheduledMsgCache.add(originalMsg.id);
              SendLater.updateStatusIndicator();
            }
          );
```

The `originalMessage` is not a MessageHeader but a MessagePart. It does not have an `id` member, so the four lines manipulating the MsgCache do not seem to do anything useful. 

Why is the manipulation repeated after the message has been saved? That seems not necessary, is it?

Reply:
I updated the PR and now use the correct msgId and removed the superfluous update of the *scheduledMsgCache maps.

Reply:
I think it would improve performance slightly to put back the code that updates the two caches after the message is saved, but using the correct id for the newly saved draft rather than the old id. I don't think it hurts too much not to do it -- Send Later will just pick up that draft in the next scan and add it to the cache in `possiblySendMessage`.

I know this is probably out-of-scope for this PR, but it's silly that there are two different caches. There should be just be one cache, a JavaScript dictionary whose keys are message IDs and values are either true for scheduled or false for unscheduled.

Also, as far as I can tell, right now messages that have been deleted aren't ever being removed from the caches, so that's a memory leak if TB is kept running for a long time.

Reply:
I really do not understand that piece of code. This is the current version (removed unimportant parts):

```
SLTools.scheduledMsgCache.delete(originalMsg.id);
SLTools.unscheduledMsgCache.add(originalMsg.id);

// Check if original message has x-send-later headers
if (originalMsg.headers.hasOwnProperty("x-send-later-at")) {
  let { preferences, scheduleCache } = await messenger.storage.local.get({ preferences: {}, scheduleCache: {} });

  // Re-save message (drops x-send-later headers by default
  // because they are not loaded when editing as draft).
  messenger.compose.saveMessage(tab.id, {mode: "draft"}).then(
	() => {
	  SLTools.scheduledMsgCache.delete(originalMsg.id);
	  SLTools.unscheduledMsgCache.add(originalMsg.id);
	  SendLater.updateStatusIndicator();
	}
  );
````

So in the first two lines we do:

```
SLTools.scheduledMsgCache.delete(originalMsg.id);
SLTools.unscheduledMsgCache.add(originalMsg.id);
```

and in the then() callback we do again:

```
SLTools.scheduledMsgCache.delete(originalMsg.id);
SLTools.unscheduledMsgCache.add(originalMsg.id);
SendLater.updateStatusIndicator();
```

That is exactly the same, the cache is not reverted or anything in the second step, we just delete the same entry from the `scheduledMsgCache` map (which is not there anymore = no op) and we add the same entry to the `unscheduledMsgCache` map (which is already there = no op).

It really is not doing anything. Or am I missing something? We definitely should elaborate, what these caches are used for and try to simplify them.

Reply:
@jperryhouts : Do you have time to help us understand what these caches are for?